File: Maze.scala
/* Maze generator in Scala
 * Joe Wingbermuehle
 * 2010-10-16
 */

object Maze {

   private var width:   Int = 0
   private var height:  Int = 0
   private var maze:    Array[Array[Int]] = null

   /* Show the maze. */
   def show {
      maze.foreach(row =>
         {
            row.foreach(block => print(if(block == 1) "[]" else "  "))
            println
         }
      )
   }

   /* Start carving at x, y. */
   private def carve(x: Int, y: Int) {

      def update_pos(dir: Int, x: Int, y: Int): (Int, Int) = dir match {
         case 0 => (x + 1, y + 0)
         case 1 => (x + 0, y + 1)
         case 2 => (x - 1, y + 0)
         case _ => (x + 0, y - 1)
      }

      var dir:    Int = (scala.math.random * 4.0).toInt
      var count:  Int = 0
      while(count < 4) {
         val (x1, y1) = update_pos(dir, x, y)
         val (x2, y2) = update_pos(dir, x1, y1)
         if(x2 > 0 && x2 < width && y2 > 0 && y2 < height) {
            if(maze(y1)(x1) == 1 && maze(y2)(x2) == 1) {
               maze(y1)(x1) = 0
               maze(y2)(x2) = 0
               carve(x2, y2)
            }
         }
         count += 1
         dir = (dir + 1) % 4
      }

   }

   /* Generate a maze. */
   def generate(w: Int, h: Int) {
      width = w
      height = h
      maze = Array.fill[Int](height, width)(1)
      maze(1)(1) = 0
      carve(1, 1)
      maze(0)(1) = 0
      maze(height - 1)(width - 2) = 0
   }

   /* Generate and display a random maze. */
   def main(args: Array[String]) {
      generate(39, 23)
      show
   }

}

File: maze.rb
#!/usr/bin/ruby
# Maze generator in Ruby
# Joe Wingbermuehle
# 2011-06-07

class Maze

   # Generate a maze.
   def generate(width, height)
      @width = width
      @height = height
      @maze = Array.new
      for y in 0 ... @height
         for x in 0 ... @width
            @maze[y * @width + x] = 1
         end
      end
      @maze[1 * @width + 1] = 0
      carve(1, 1)
      @maze[0 * @width + 1] = 0
      @maze[(@height - 1) * @width + @width - 2] = 0
   end

   # Show the maze.
   def show
      for y in 0 ... @height
         for x in 0 ... @width
            chars = if @maze[y * @width + x] == 1 then
               [ "[", "]" ]
            else
               [ " ", " " ]
            end
            chars.each { |c| putc(c) }
         end
         puts
      end
   end

   private

   # Start carving at x, y.
   def carve(x, y)

      def update_pos(dir, x, y)
         case dir
         when 0
            return x + 1, y + 0
         when 1
            return x + 0, y + 1
         when 2
            return x - 1, y + 0
         when 3
            return x + 0, y - 1
         end
      end

      dir = rand(4)
      count = 0
      while count < 4
         x1, y1 = update_pos(dir, x, y)
         x2, y2 = update_pos(dir, x1, y1)
         if x2 > 0 and x2 < @width and y2 > 0 and y2 < @height
            if @maze[y1 * @width + x1] == 1 and @maze[y2 * @width + x2] == 1
               @maze[y1 * @width + x1] = 0
               @maze[y2 * @width + x2] = 0
               carve(x2, y2)
            end
         end
         count += 1
         dir = (dir + 1) % 4
      end

   end

end

mz = Maze.new
mz.generate(39, 23)
mz.show()

